File: test/board/GetFlipTest.java
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package board;

import bitpattern.BitPattern;
import java.util.ArrayList;
import java.util.Arrays;
import org.junit.Test;
import static org.junit.Assert.*;

/**
 *
 * @author michele
 */
public class GetFlipTest {
  GetFlip FLIPPER = GetFlip.load();
  
  @Test
  public void testFlip() {
    Board board = new Board();
    ArrayList<Long> possibleMoves = new ArrayList<>(Arrays.asList(
            BitPattern.parsePattern("--------\n"
                                  + "--------\n" 
                                  + "---X----\n"
                                  + "---X----\n"
                                  + "--------\n"
                                  + "--------\n"
                                  + "--------\n"
                                  + "--------\n"),
            BitPattern.parsePattern("--------\n"
                                  + "--------\n" 
                                  + "--------\n"
                                  + "--XX----\n"
                                  + "--------\n"
                                  + "--------\n"
                                  + "--------\n"
                                  + "--------\n"),
            BitPattern.parsePattern("--------\n"
                                  + "--------\n" 
                                  + "--------\n"
                                  + "--------\n"
                                  + "----XX--\n"
                                  + "--------\n"
                                  + "--------\n"
                                  + "--------\n"),
            BitPattern.parsePattern("--------\n"
                                  + "--------\n" 
                                  + "--------\n"
                                  + "--------\n"
                                  + "----X---\n"
                                  + "----X---\n"
                                  + "--------\n"
                                  + "--------\n")));

    for (long flip : possibleMoves) {
      int move = Long.numberOfTrailingZeros(flip & ~(board.getPlayer() | board.getOpponent()));
      long newFlip = FLIPPER.getFlip(move, board.getPlayer(), board.getOpponent());
      if (flip != newFlip) {
        System.out.println(board);
        System.out.println(BitPattern.patternToString(flip));
        System.out.println(BitPattern.patternToString(newFlip));
        assert(false);
      }
    }
    assertEquals(0, FLIPPER.getFlip(0, board.getPlayer(), board.getOpponent()));
  }
  

  @Test
  public void testFlipManyDir() {
    PossibleMovesFinderImproved movesFinder = new PossibleMovesFinderImproved();
    Board board = new Board("--------\n"
                          + "-O-O----\n" 
                          + "--XX--O-\n"
                          + "---X-X--\n"
                          + "---XX---\n"
                          + "--------\n"
                          + "--------\n"
                          + "--------\n", false);
    ArrayList<Long> possibleMoves = new ArrayList<>(Arrays.asList(
            BitPattern.parsePattern("--------\n"
                             + "--------\n" 
                             + "---X----\n"
                             + "---X-X--\n"
                             + "---XX---\n"
                             + "---X----\n"
                             + "--------\n"
                             + "--------\n"),
            BitPattern.parsePattern("--------\n"
                             + "--------\n" 
                             + "--X-----\n"
                             + "---X----\n"
                             + "----X---\n"
                             + "-----X--\n"
                             + "--------\n"
                             + "--------\n"),
            BitPattern.parsePattern("--------\n"
                             + "--------\n" 
                             + "--X-----\n"
                             + "-X------\n"
                             + "--------\n"
                             + "--------\n"
                             + "--------\n"
                             + "--------\n")));

    for (long flip : possibleMoves) {
      int move = Long.numberOfTrailingZeros(flip & ~(board.getPlayer() | board.getOpponent()));
      long newFlip = FLIPPER.getFlip(move, board.getPlayer(), board.getOpponent());
      if (flip != newFlip) {
        System.out.println(board);
        System.out.println(BitPattern.patternToString(flip));
        System.out.println(BitPattern.patternToString(newFlip));
        assert(false);
      }
    }
  }
  
  @Test
  public void testFindPossibleMovesRandom() {
    PossibleMovesFinderImproved movesFinder = new PossibleMovesFinderImproved();
    int n = 100000;
    boolean movesFound[] = new boolean[64];
 
    for (int i = 0; i < n; i++) {
      Arrays.fill(movesFound, false);

      Board board = Board.randomBoard();
      long[] flips = movesFinder.possibleMoves(board);

      for (long flip : flips) {
        int move = Long.numberOfTrailingZeros(flip & ~(board.getPlayer() | board.getOpponent()));
        long newFlip = 0;
        if (board.getEmptySquares() == 1) {
          newFlip = FLIPPER.getFlipLast(move, board.getOpponent());
          newFlip = newFlip & ~board.getPlayer();
        } else {
          newFlip = FLIPPER.getFlip(move, board.getPlayer(), board.getOpponent());          
        }
        if (flip != newFlip) {
          System.out.println(board);
          System.out.println(board.getEmptySquares());
          System.out.println(BitPattern.patternToString(flip));
          System.out.println(BitPattern.patternToString(newFlip));
          assert(false);
        }
        movesFound[move] = true;
      }
      for (int move = 0; move < movesFound.length; ++move) {
        if (!movesFound[move]) {
          if (FLIPPER.getFlip(move, board.getPlayer(), board.getOpponent()) != 0) {
            System.out.println(board);
            System.out.println(move);
            assert(false);
          }
        }
      }
    }
  }
}

File: app/src/main/cpp/get_flip_test.cpp
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <iostream>
#include <string>
#include "bitpattern.h"
#include "board.h"
#include "get_flip.h"

TEST(GetFlip, Basic) {
  Board horiz("--------"
              "------OX"
              "--------"
              "-OOOOOOX"
              "XOOOOOO-"
              "-----XOO"
              "--------"
              "OOX-----");
  BitPattern flip1 = ParsePattern("--------"
                                  "-----XX-"
                                  "--------"
                                  "--------"
                                  "--------"
                                  "--------"
                                  "--------"
                                  "--------");
  BitPattern flip2 = ParsePattern("--------"
                                  "--------"
                                  "--------"
                                  "XXXXXXX-"
                                  "--------"
                                  "--------"
                                  "--------"
                                  "--------");
  BitPattern flip3 = ParsePattern("--------"
                                  "--------"
                                  "--------"
                                  "--------"
                                  "-XXXXXXX"
                                  "--------"
                                  "--------"
                                  "--------");
  
  if (GetFlipBasic(50, horiz.GetPlayer(), horiz.GetOpponent()) != flip1) {
    std::cout << PatternToString(flip1) << "\n";
    std::cout << PatternToString(GetFlipBasic(50, horiz.GetPlayer(), horiz.GetOpponent()));
  }
  if (GetFlipBasic(39, horiz.GetPlayer(), horiz.GetOpponent()) != flip2) {
    std::cout << PatternToString(flip2) << "\n";
    std::cout << PatternToString(GetFlipBasic(39, horiz.GetPlayer(), horiz.GetOpponent()));
  }
  if (GetFlipBasic(24, horiz.GetPlayer(), horiz.GetOpponent()) != flip3) {
    std::cout << PatternToString(flip2) << "\n";
    std::cout << PatternToString(GetFlipBasic(24, horiz.GetPlayer(), horiz.GetOpponent()));
  }
  if (GetFlipBasic(24, horiz.GetPlayer(), horiz.GetOpponent()) != flip3) {
    std::cout << PatternToString(flip2) << "\n";
    std::cout << PatternToString(GetFlipBasic(24, horiz.GetPlayer(), horiz.GetOpponent()));
  }

  EXPECT_EQ(GetFlipBasic(8, horiz.GetPlayer(), horiz.GetOpponent()), 0);
  EXPECT_EQ(GetFlipBasic(15, horiz.GetPlayer(), horiz.GetOpponent()), 0);
}

TEST(GetFlip, Outflank) {
  int flip = 0b00111010;
  EXPECT_EQ(kOutflank[0 << 8 | flip], 0b00000100);
  EXPECT_EQ(kOutflank[1 << 8 | flip], 0);
  EXPECT_EQ(kOutflank[2 << 8 | flip], 0b01000001);
  EXPECT_EQ(kOutflank[4 << 8 | flip], 0);
  EXPECT_EQ(kOutflank[6 << 8 | flip], 0b00000100);

  EXPECT_EQ(kFlip[0 << 8 | 0b00001000], 0b00000111);
  EXPECT_EQ(kFlip[1 << 8 | 0b00001000], 0b00000110);
  EXPECT_EQ(kFlip[2 << 8 | 0b00001000], 0);
  EXPECT_EQ(kFlip[3 << 8 | 0b00001000], 0);
  EXPECT_EQ(kFlip[4 << 8 | 0b00001000], 0);
  EXPECT_EQ(kFlip[5 << 8 | 0b00001000], 0b00110000);
  EXPECT_EQ(kFlip[6 << 8 | 0b00001000], 0b01110000);
  EXPECT_EQ(kFlip[7 << 8 | 0b00001000], 0b11110000);

  EXPECT_EQ(kFlip[0 << 8 | 0b01000010], 0);
  EXPECT_EQ(kFlip[1 << 8 | 0b01000010], 0);
  EXPECT_EQ(kFlip[2 << 8 | 0b01000010], 0);
  EXPECT_EQ(kFlip[3 << 8 | 0b01000010], 0b00111100);
  EXPECT_EQ(kFlip[4 << 8 | 0b01000010], 0b00111100);
  EXPECT_EQ(kFlip[5 << 8 | 0b01000010], 0);
  EXPECT_EQ(kFlip[6 << 8 | 0b01000010], 0);
  EXPECT_EQ(kFlip[7 << 8 | 0b01000010], 0);
}

TEST(GetFlip, Random) {
  for (int i = 0; i < 1000000; ++i) {
    Board b = RandomBoard();
    Move move = rand() % 64;
    if (b.IsEmpty(move)) {
      BitPattern flip_basic = GetFlipBasic(move, b.GetPlayer(), b.GetOpponent());
      BitPattern flip = GetFlip(move, b.GetPlayer(), b.GetOpponent());
      if (flip_basic != flip) {
        std::cout << "Error\n";
        std::cout << b.ToString() << " " << move << "\n";
        std::cout << PatternToString(flip_basic) << "\n";
        std::cout << PatternToString(flip) << "\n";
      }
      EXPECT_EQ(flip_basic, flip);
    }
  }
  PrintMoveMetadata();
}

